Converted shutdown.js test to selenium python test

emiliotl opened the pull request:
Contribution by converting the shutdown test from js to py with selenium. Solving part of #3335

minrk commented on notebook/tests/selenium/test_shutdown.py:
```diff
@@ -0,0 +1,13 @@
+    notebook.browser.find_element_by_id('shutdown_kernel').click()
+    notebook.browser.find_element_by_class_name('btn.btn-default.btn-sm.btn-danger').click()
+
+    notebook.wait_explicitly(3)
```
Is it possible to do this without a time.sleep? Instead, perhaps wait for a condition? Sleep with a timer leaves a race condition in slow environments.

emiliotl replied:
Sure, you're right. Instead of `time.sleep` I will create a method which uses one of the wait ones, so its time to wait can be flexible. I will send the fix, thank you for reviewed it 👍

emiliotl replied:
@minrk let me know if it needs more refactor, please : D

takluyver commented on notebook/tests/selenium/test_shutdown.py:
```diff
@@ -0,0 +1,13 @@
+    notebook.execute_cell(0)
+
+    assert not notebook.is_kernel_running()
+    assert notebook.get_cell_output() == None
```
Does this test actually fail if you comment out the lines to shut the kernel down? I.e. does the output normally appear quickly enough that it would be there before Selenium looks for it?

emiliotl replied:
@takluyver I commented the kernel shutdown statements and its corresponding assertion, and the test failed because the output was generated.

I didn't add an implicit or explicit wait since it could lead more to a race condition. And I could use something like `wait_for_element_availability` but that method depends of the output element to actually appear, which is something we don't expect since the kernel is shutdown.

takluyver replied:
Yup, no problem that it's not properly waiting - I don't know of any good way to test that something *doesn't* happen asynchronously. So long as it's reasonably likely to fail when the relevant lines are commented out, it's still a useful test.

takluyver commented on notebook/tests/selenium/utils.py:
```diff
@@ -195,13 +195,22 @@ def wait_for_stale_cell(self, cell):
 
+    def get_cell_output(self, index=0, output='output_subarea'):
+        try:
+            return self.cells[index].find_element_by_class_name(output).text
```
This is singular, so it's only going to be the first output of the cell, which may have multiple outputs. Maybe we should call it `get_cell_outputs`, and return whatever output elements there are.

emiliotl replied:
@takluyver sorry, I don't understand the multiple outputs part. Do you mean for an output that has several lines? Like using a loop that prints the value line by line? The point of the method is to get the whole text on the output area of a single cell. So, if we put a loop on the cell, the output, even if there are several lines, will be captured, the detail would be that the result would be a string with `\n` as breaklines. If someone would like to work on that output separately, a split method would be required. 

Having said that, I may have misunderstood your perspective 😅, so tell me if I'm far from what you commented lol

takluyver replied:
Conceptually, in a notebook, each cell can have zero-to-many outputs, each of which may be one of various different types, e.g. plain text, HTML, or an image. See here for more info: https://nbformat.readthedocs.io/en/latest/format_description.html#code-cell-outputs

So I'm suggesting that rather than returning text, this method should return a list of browser elements for the outputs from that cell. That also conveniently means we don't need a special case `None` for when there's no output - it can just return an empty list then.

emiliotl commented:
After seeing the reviews on this PR, I executed the tests several times to check some of the issues mentioned, and I notice that there is a race condition, making the results not consistent. I'll change the title to [WIP] and I will notify both reviewers once I got it solved. Sorry for the trouble.

takluyver commented:
No worries, thanks for picking up the race condition before the PR was merged.

emiliotl commented:
@takluyver @minrk I solved the race condition and improved the find methods to cover that issue as well. Also, I changed the `get_cell_output` to get the selenium object instead of just the text. I run the test several times and all passed, so I removed the WIP on the title so it can be reviewed. I see the AppVeyor test failed for some npm install, if there is something wrong with the PR, please let me know. Thank you for all the comments 👍

takluyver commented:
Thanks, looks good to me.
